File: plugins/plugin-msteams/typescript/src/actions/sendMessage.ts
import type {
  Action,
  ActionExample,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { MSTEAMS_SERVICE_NAME, type MSTeamsService } from "../service";

export const SEND_MESSAGE_ACTION = "SEND_MSTEAMS_MESSAGE";

export const sendMessageAction: Action = {
  name: SEND_MESSAGE_ACTION,
  similes: [
    "MSTEAMS_SEND_MESSAGE",
    "MSTEAMS_REPLY",
    "MSTEAMS_MESSAGE",
    "SEND_TEAMS_MESSAGE",
    "REPLY_TEAMS",
    "TEAMS_SEND",
    "TEAMS_MESSAGE",
  ],
  description: "Send a message to a Microsoft Teams conversation",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
  ): Promise<boolean> => {
    const source = message.content?.source;
    return source === "msteams";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const msTeamsService = runtime.getService(
      MSTEAMS_SERVICE_NAME,
    ) as unknown as MSTeamsService | undefined;
    if (!msTeamsService) {
      if (callback) {
        await callback({
          text: "MS Teams service not available",
        });
      }
      return { success: false, error: "MS Teams service not initialized" };
    }

    const client = msTeamsService.getClient();
    if (!client) {
      if (callback) {
        await callback({
          text: "MS Teams client not available",
        });
      }
      return { success: false, error: "MS Teams client not initialized" };
    }

    const currentState = state ?? (await runtime.composeState(message));
    const responseText = currentState.values?.response?.toString() || "";
    const conversationId = message.content?.conversationId as
      | string
      | undefined;

    if (!conversationId) {
      if (callback) {
        await callback({
          text: "No conversation ID available",
        });
      }
      return { success: false, error: "Missing conversation ID" };
    }

    try {
      const result = await client.sendProactiveMessage(
        conversationId,
        responseText,
      );

      if (callback) {
        await callback({
          text: responseText,
          action: SEND_MESSAGE_ACTION,
        });
      }

      return {
        success: true,
        data: {
          action: SEND_MESSAGE_ACTION,
          conversationId,
          messageId: result.messageId,
          text: responseText,
        },
      };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);

      if (callback) {
        await callback({
          text: `Failed to send message: ${errorMessage}`,
        });
      }

      return { success: false, error: errorMessage };
    }
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Send a message to this Teams conversation",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll send a message to this Teams chat now.",
          actions: [SEND_MESSAGE_ACTION],
        },
      },
    ],
  ] as ActionExample[][],
};

export const SEND_POLL_ACTION = "SEND_MSTEAMS_POLL";

export const sendPollAction: Action = {
  name: SEND_POLL_ACTION,
  similes: [
    "MSTEAMS_SEND_POLL",
    "MSTEAMS_CREATE_POLL",
    "TEAMS_POLL",
    "CREATE_POLL",
    "SEND_POLL",
  ],
  description: "Send a poll to a Microsoft Teams conversation",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
  ): Promise<boolean> => {
    const source = message.content?.source;
    return source === "msteams";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const msTeamsService = runtime.getService(
      MSTEAMS_SERVICE_NAME,
    ) as unknown as MSTeamsService | undefined;
    if (!msTeamsService) {
      if (callback) {
        await callback({
          text: "MS Teams service not available",
        });
      }
      return { success: false, error: "MS Teams service not initialized" };
    }

    const client = msTeamsService.getClient();
    if (!client) {
      if (callback) {
        await callback({
          text: "MS Teams client not available",
        });
      }
      return { success: false, error: "MS Teams client not initialized" };
    }

    const currentState = state ?? (await runtime.composeState(message));
    const conversationId = message.content?.conversationId as
      | string
      | undefined;
    const question = currentState.values?.pollQuestion?.toString() || "";
    const optionsRaw = currentState.values?.pollOptions;

    if (!conversationId) {
      return { success: false, error: "Missing conversation ID" };
    }

    if (!question) {
      return { success: false, error: "Missing poll question" };
    }

    // Parse options
    let options: string[] = [];
    if (Array.isArray(optionsRaw)) {
      options = optionsRaw.map((o) => String(o));
    } else if (typeof optionsRaw === "string") {
      options = optionsRaw.split(",").map((o) => o.trim());
    }

    if (options.length < 2) {
      return { success: false, error: "Poll must have at least 2 options" };
    }

    try {
      const result = await client.sendPoll(conversationId, question, options);

      if (callback) {
        await callback({
          text: `Poll created: ${question}`,
          action: SEND_POLL_ACTION,
        });
      }

      return {
        success: true,
        data: {
          action: SEND_POLL_ACTION,
          conversationId,
          pollId: result.pollId,
          messageId: result.messageId,
          question,
          options,
        },
      };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      return { success: false, error: errorMessage };
    }
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Create a poll asking what day works best for our meeting",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll create a poll for that.",
          actions: [SEND_POLL_ACTION],
        },
      },
    ],
  ] as ActionExample[][],
};

export const SEND_ADAPTIVE_CARD_ACTION = "SEND_MSTEAMS_CARD";

export const sendAdaptiveCardAction: Action = {
  name: SEND_ADAPTIVE_CARD_ACTION,
  similes: [
    "MSTEAMS_SEND_CARD",
    "MSTEAMS_ADAPTIVE_CARD",
    "TEAMS_CARD",
    "SEND_CARD",
  ],
  description: "Send an Adaptive Card to a Microsoft Teams conversation",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
  ): Promise<boolean> => {
    const source = message.content?.source;
    return source === "msteams";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State | undefined,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const msTeamsService = runtime.getService(
      MSTEAMS_SERVICE_NAME,
    ) as unknown as MSTeamsService | undefined;
    if (!msTeamsService) {
      return { success: false, error: "MS Teams service not initialized" };
    }

    const client = msTeamsService.getClient();
    if (!client) {
      return { success: false, error: "MS Teams client not initialized" };
    }

    const currentState = state ?? (await runtime.composeState(message));
    const conversationId = message.content?.conversationId as
      | string
      | undefined;
    const cardContent = currentState.values?.cardContent as
      | Record<string, unknown>
      | undefined;

    if (!conversationId) {
      return { success: false, error: "Missing conversation ID" };
    }

    if (!cardContent) {
      return { success: false, error: "Missing card content" };
    }

    try {
      const cardBody = Array.isArray(cardContent.body) ? cardContent.body : [];
      const cardActions = Array.isArray(cardContent.actions)
        ? cardContent.actions
        : undefined;
      const card = {
        type: "AdaptiveCard" as const,
        version: "1.5",
        body: cardBody as unknown[],
        actions: cardActions as unknown[] | undefined,
      };

      const result = await client.sendAdaptiveCard(
        conversationId,
        card,
        currentState.values?.fallbackText?.toString(),
      );

      if (callback) {
        await callback({
          text: "Adaptive Card sent",
          action: SEND_ADAPTIVE_CARD_ACTION,
        });
      }

      return {
        success: true,
        data: {
          action: SEND_ADAPTIVE_CARD_ACTION,
          conversationId,
          messageId: result.messageId,
        },
      };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      return { success: false, error: errorMessage };
    }
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Send a card with meeting details",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll send an Adaptive Card with the meeting information.",
          actions: [SEND_ADAPTIVE_CARD_ACTION],
        },
      },
    ],
  ] as ActionExample[][],
};

File: plugins/plugin-msteams/rust/src/actions/send_message.rs
//! Send message actions for MS Teams.

use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, MSTeamsAction};
use crate::error::Result;

/// Action that sends a message to an MS Teams conversation.
pub struct SendMessageAction;

#[async_trait]
impl MSTeamsAction for SendMessageAction {
    fn name(&self) -> &'static str {
        "SEND_MSTEAMS_MESSAGE"
    }

    fn description(&self) -> &'static str {
        "Send a message to a Microsoft Teams conversation"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context.message.get("source").and_then(|v| v.as_str());
        Ok(source == Some("msteams"))
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let response_text = context
            .state
            .get("response")
            .and_then(|v| v.get("text"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        Ok(serde_json::json!({
            "action": self.name(),
            "conversationId": context.conversation_id,
            "text": response_text,
            "replyToId": context.message.get("activityId")
        }))
    }
}

/// Action that sends a poll to an MS Teams conversation.
pub struct SendPollAction;

#[async_trait]
impl MSTeamsAction for SendPollAction {
    fn name(&self) -> &'static str {
        "SEND_MSTEAMS_POLL"
    }

    fn description(&self) -> &'static str {
        "Send a poll to a Microsoft Teams conversation"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context.message.get("source").and_then(|v| v.as_str());
        Ok(source == Some("msteams"))
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let question = context
            .state
            .get("pollQuestion")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let options = context
            .state
            .get("pollOptions")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str())
                    .map(|s| s.to_string())
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default();

        let max_selections = context
            .state
            .get("maxSelections")
            .and_then(|v| v.as_u64())
            .unwrap_or(1) as u32;

        Ok(serde_json::json!({
            "action": self.name(),
            "conversationId": context.conversation_id,
            "question": question,
            "options": options,
            "maxSelections": max_selections
        }))
    }
}

/// Action that sends an Adaptive Card to an MS Teams conversation.
pub struct SendAdaptiveCardAction;

#[async_trait]
impl MSTeamsAction for SendAdaptiveCardAction {
    fn name(&self) -> &'static str {
        "SEND_MSTEAMS_CARD"
    }

    fn description(&self) -> &'static str {
        "Send an Adaptive Card to a Microsoft Teams conversation"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context.message.get("source").and_then(|v| v.as_str());
        Ok(source == Some("msteams"))
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let card = context
            .state
            .get("cardContent")
            .cloned()
            .unwrap_or(serde_json::Value::Null);

        let fallback_text = context
            .state
            .get("fallbackText")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        Ok(serde_json::json!({
            "action": self.name(),
            "conversationId": context.conversation_id,
            "card": card,
            "fallbackText": fallback_text
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_msteams_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "msteams",
                "text": "Hello"
            }),
            conversation_id: "conv-123".to_string(),
            user_id: "user-456".to_string(),
            tenant_id: Some("tenant-789".to_string()),
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_non_msteams_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "discord",
                "text": "Hello"
            }),
            conversation_id: "conv-123".to_string(),
            user_id: "user-456".to_string(),
            tenant_id: None,
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_send_poll_action() {
        let action = SendPollAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "msteams"
            }),
            conversation_id: "conv-123".to_string(),
            user_id: "user-456".to_string(),
            tenant_id: None,
            state: serde_json::json!({
                "pollQuestion": "What's your favorite color?",
                "pollOptions": ["Red", "Blue", "Green"],
                "maxSelections": 1
            }),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["action"], "SEND_MSTEAMS_POLL");
        assert_eq!(result["question"], "What's your favorite color?");
    }
}
